packages/client/src/client/Web/serviceWorkerRegistration.ts

```diff
@@ -53,51 +56,78 @@ export function register(config?: Config) {
         })
       } else {
         // Is not localhost. Just register service worker
-        registerValidSW(swUrl, config)
+        registerServiceWorker(swUrl, config)
       }
     })
   }
 }
 
-function registerValidSW(swUrl: string, config?: Config) {
+function addStateChangeHandler(
+  serviceWorkerInstance: ServiceWorker,
+  registration: ServiceWorkerRegistration,
+  registrationConfig: Config | undefined,
+) {
+  serviceWorkerInstance.onstatechange = () => {
+    console.debug(
+      "'installing' SW changed state to: ",
+      serviceWorkerInstance.state,
+    )
+    if (serviceWorkerInstance.state === "installed") {
+      if (navigator.serviceWorker.controller) {
+        // At this point, the updated precached content has been fetched,
+        // but the previous service worker will still serve the older
+        // content until all client tabs are closed.
+        console.debug(
+          "New content is available and will be used when all " +
+            "tabs for this page are closed. See https://cra.link/PWA.",
+        )
+        serviceWorkerInstance.postMessage({ type: "SKIP_WAITING" })
+      } else {
+        // At this point, everything has been precached.
+        // It's the perfect time to display a
+        // "Content is cached for offline use." message.
+        console.debug("Content is cached for offline use.")
+
+        // Execute callback
+        if (registrationConfig && registrationConfig.onSuccess) {
```

Comment:
is the `onSuccess`  redundant? I don't see where the callback is defined.

Reply:
yes, I think it is .. some of this I think is inherited from CRA .. I am not worried about it, I kind of like the fact that is obviously CRA so we can use that as a reference, until we completely rewrite it as a custom thing .. which may never happen